packages/@lwc/style-compiler/src/serialize.ts

```diff
@@ -56,10 +57,29 @@ export default function serialize(result: Result, config: Config): string {
     const serializedStyle = serializeCss(result, collectVarFunctions).trim();
 
     if (serializedStyle) {
-        // inline function
-        buffer += `function stylesheet(${HOST_SELECTOR_IDENTIFIER}, ${SHADOW_SELECTOR_IDENTIFIER}, ${SHADOW_DOM_ENABLED_IDENTIFIER}) {\n`;
-        buffer += `  return ${serializedStyle};\n`;
-        buffer += `}\n`;
+        // In the case where adopted stylesheets (aka construtable stylesheets) are supported,
+        // we use a fast path where the CSSStyleSheet object is cached. This avoids having to use
+        // some kind of mapping from strings to CSSStyleSheet objects somewhere, since the whole point
+        // of adoptedStyleSheets is to avoid duplication. Also with native shadow, the string never changes,
+        // so there is no point running stylesheet() over and over again to do costly string concatenation,
+        // so that's another benefit of caching the CSSStyleSheet object.
+        buffer += `\n
+var cachedStylesheet
+
+function generateCss(${HOST_SELECTOR_IDENTIFIER}, ${SHADOW_SELECTOR_IDENTIFIER}, ${SHADOW_DOM_ENABLED_IDENTIFIER}) {
+  return ${serializedStyle};
+}
+
+function stylesheet(${HOST_SELECTOR_IDENTIFIER}, ${SHADOW_SELECTOR_IDENTIFIER}, ${SHADOW_DOM_ENABLED_IDENTIFIER}, ${HAS_ADOPTED_STYLESHEETS_IDENTIFIER}) {
```

Comment:
It would be interesting to move this into a helper in `@lwc/engine-core` to avoid code duplication.

Reply:
I'm not quite sure how to do that; it seems the `import` I add is left unmodified in the final code output. https://github.com/salesforce/lwc/commit/86cd0659526eac93157f97bb0280b7ced5c7592d

If you have some advice about how to do this, that would help me a lot. :)